bigframes/ml/compose.py

```diff
@@ -46,6 +48,114 @@
 )
 
 
+CUSTOM_TRANSFORMER_SQL_RX = re.compile(
+    "^(?P<sql>.*)/[*]CT.(?P<id>[A-Z]+[A-Z0-9]*)[(](?P<config>[^*]*)[)][*]/$",
+    re.IGNORECASE,
+)
+
+
+class CustomTransformer(base.BaseTransformer):
+    _CTID = None
+    _custom_transformer_classes = {}
+
+    @classmethod
+    def register(cls, transformer_cls: Type[base.BaseTransformer]):
+        assert transformer_cls._CTID
+        cls._custom_transformer_classes[transformer_cls._CTID] = transformer_cls
+
+    @classmethod
+    def find_matching_transformer(
+        cls, transform_sql: str
+    ) -> Optional[Type[base.BaseTransformer]]:
+        for transform_cls in cls._custom_transformer_classes.values():
+            if transform_cls.understands(transform_sql):
+                return transform_cls
+        return None
+
+    @classmethod
+    def understands(cls, transform_sql: str) -> bool:
+        """
+        may be overwritten to have a more advanced matching, possibly without comments in SQL
+        """
+        m = CUSTOM_TRANSFORMER_SQL_RX.match(transform_sql)
+        if m and m.group("id").strip() == cls._CTID:
+            return True
+        return False
+
+    def __init__(self):
+        super().__init__()
+
+    def _compile_to_sql(
+        self, X: bpd.DataFrame, columns: Optional[Iterable[str]] = None
+    ) -> List[str]:
+        if columns is None:
+            columns = X.columns
+        return [
+            f"{self.custom_compile_to_sql(X, column)} {self._get_sql_comment(column)} AS {self.get_target_column_name(column)}"
+            for column in columns
+        ]
+
+    def get_target_column_name(self, column: str) -> str:
+        return f"{self._CTID.lower()}_{column}"
+
+    @classmethod
+    @abc.abstractclassmethod
+    def custom_compile_to_sql(cls, X: bpd.DataFrame, column: str) -> str:
```

Comment:
As much as possible, I'd like to avoid having to fallback to SQL as part of our public interfaces. I do also see the utility in providing such an extensions point, though.

If we do provide such hooks, I'd like it to be via a more general purpose API such as https://github.com/googleapis/python-bigquery-dataframes/issues/647 and have CustomTransform be able to hook into our compilation somehow to generate a TRANSFORM clause instead of a full table expression (e.g. SELECT ... FROM ...).

Reply:
just my 2 cents:

looking at the into of the pandasql README:

```
pandasql [...] provides a more familiar way of manipulating and cleaning data for people new to Python or pandas [using SQL syntax].
```

So it is thought for developers who are already familiar with SQL and new to pandas.
Our intention to use BigFrames is to please developers who are familiar with Python and pandas but do not know much about SQL.
:-)

I understand your concerns to break the BigFrames philosophie to abstract all the BQ SQL stuff behind the pandas interface. 

ColumnTransformer already is a very good abstraction of the TRANSFORM clause.
It manages very good all the "ML." transformations but lacks support for other transformations (mostly CASE WHEN).

So, BQ SQL experts can create a set of CustomTransformers and the DataScientists using BigFrames will never get in touch with BQ SQL. They use the CustomTransformers in the same way, as they use the predefined transformers.

Reply:
I see. So, the custom transformation API is targeted at a different user base. I'd still like to do some thinking to see if there's a way we can offer a SQL escape hatch in line with long term plans.